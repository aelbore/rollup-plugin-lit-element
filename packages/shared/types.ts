export type { InputPlugin } from 'qoi-cli'

export type Enforce = 'pre' | 'post'
export type ViteInject = { inject?: boolean }
export type Vite = ViteInject | boolean

export type TSConfig = { 
  overridePaths?: boolean
  compilerOptions?: { paths?: { [from: string]: [string] }  } 
}

export interface Options {
  vite?: Vite
  enforce?: Enforce
  importPackage?: string
  overridePaths?: boolean
  exclude?: string | string[]
  minify?: boolean
  sassOptions?: import('sass').Options<'sync'>
  paths?: {
    [from: string]: [string]
  }
}