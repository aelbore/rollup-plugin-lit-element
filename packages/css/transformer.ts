import { Visitor } from '@swc/core/Visitor.js'
import * as swc from 'swc-ast-helpers'

import type { 
  Module,
  ExportDefaultExpression, 
  StringLiteral, 
  VariableDeclaration, 
  VariableDeclarator, 
  Expression,
  ModuleItem
} from '@swc/core'

import type { Options } from '../shared'

const isVariableDeclarator = (declaration: VariableDeclarator, value: string) => 
  swc.isVariableDeclarator(declaration) 
    && swc.isIdentifer(declaration.id)
    && declaration.id.value.includes(value)
    && swc.isStringLiteral(declaration.init as Expression)

const isVariableDeclaration = (value: string) => {
  return (item: ModuleItem) => 
    swc.isVariableDeclaration(item) 
    && item.declarations.find(declaration => isVariableDeclarator(declaration, value))
}

export class Transformer extends Visitor {
  constructor(private options?: Options) {
    super()
  }

  visitModule(m: Module) {
    const defaultExport = m.body.find(item => swc.isExportDefaultExpression(item)) as ExportDefaultExpression

    if (defaultExport) {
      const { importPackage = 'lit' } = this.options || {}
      const cssTag = swc.createIdentifer('cssTag')

      const createTaggedTemplate = (value: string) => 
        swc.createTaggedTemplateExpression(
          cssTag, 
          swc.createTemplateLiteral([ swc.createTemplateElement(value, true) ]))

      if (swc.isStringLiteral(defaultExport.expression) || swc.isIdentifer(defaultExport.expression)) {
        const value = defaultExport.expression.value
        const item = m.body.find(isVariableDeclaration(value)) as VariableDeclaration
  
        if (item) {
          const index = m.body.indexOf(item)

          item.declarations = item.declarations.map(declaration => {
            if (isVariableDeclarator(declaration, value)) {
              const rawValue = (declaration.init as StringLiteral).value
              declaration.init = createTaggedTemplate(rawValue)
            }
            return declaration
          })
          m.body[index] = item
        }
  
        if (!item) {
          m.body.forEach(item => {
            if (swc.isExportDefaultExpression(item)) {
              const rawValue = (item.expression as StringLiteral).value
              item.expression = createTaggedTemplate(rawValue)
            }
          })
        }
      }

      m.body.forEach(item => {
        if (
          swc.isExportDefaultExpression(item) 
          && swc.isCallExpression(item.expression)
          && swc.isParenthesisExpression(item.expression.callee)
          && swc.isArrowFunctionExpression(item.expression.callee.expression)
          && swc.isStringLiteral(item.expression.callee.expression.body)
        ) {
          const rawValue = item.expression.callee.expression.body.value
          item.expression = createTaggedTemplate(rawValue)
        }
      })

      m.body.unshift(
        swc.createImportDeclaration(
          [ swc.createNamedImportSpecifier(cssTag, 'css') ], 
          swc.createStringLiteral(importPackage)
        ))
    }

    return m
  }
}

export class InlineTransformer extends Visitor {
  visitModule(m: Module) {
    m.body.forEach(item => {
      if (swc.isImportDeclaration(item) && (item.source.value.includes('.css') || item.source.value.includes('.scss'))) {
        item.source = swc.createStringLiteral(item.source.value + '?inline')
      }
    })
    return m
  }
}
