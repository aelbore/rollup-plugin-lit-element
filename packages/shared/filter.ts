import { createFilter } from '@rollup/pluginutils'

const exclude = [ /node_modules/, /virtual/ ]

export const cssfilter = createFilter('**/*.{css,scss,css?inline,scss?inline}', exclude)
export const tsJsFilter = createFilter('**/*.{js,ts,tsx,jsx,mjs}', exclude)