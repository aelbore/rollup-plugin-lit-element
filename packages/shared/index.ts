export * from './swc'
export * from './utils'
export * from './types'
export * from './filter'