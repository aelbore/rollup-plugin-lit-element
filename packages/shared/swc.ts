import { transformSync, plugins as Plugins, Plugin } from '@swc/core'

export interface SwcOptions {
  plugins?: Plugin[]
  paths?: {
    [from: string]: [string];
  }
}

export const swcTransformer = (
  code: string, 
  id: string, 
  options?: SwcOptions
) => {
  const { paths, plugins } = options || {}
  return transformSync(code, {
    jsc: {
      parser: { 
        syntax: 'typescript', 
        decorators: true, 
        dynamicImport: true 
      },
      target: 'es2022',
      ...(paths ? { paths  }: {})
    },
    filename: id,
    sourceMaps: true,
    isModule: true,
    plugin: Plugins(plugins || [])
  })
}