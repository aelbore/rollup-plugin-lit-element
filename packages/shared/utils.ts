import type { ModuleItem, ExportDeclaration, ImportDefaultSpecifier, ImportDeclaration, ClassDeclaration } from '@swc/core'
import * as swc from 'swc-ast-helpers'

export function getClassDeclaration(items: ModuleItem[]) {
  const exportDeclaration = items.find(content => swc.isExportDeclaration(content) && swc.isClasDeclaration(content.declaration))
  return (
    exportDeclaration
      ? (exportDeclaration as ExportDeclaration).declaration
      : items.find(content => swc.isClasDeclaration(content))
  ) as ClassDeclaration
}

export function hasImportDefaultSpecifierStyle(item: ModuleItem) {
  return (swc.isImportDeclaration(item) 
    && (item.specifiers.length > 0) 
    && (item.source.value.includes('.css') || item.source.value.includes('.scss')))
}

export function getStyleImportDefaultDeclarations(items: ModuleItem[]) {
  return items.reduce((p: ImportDefaultSpecifier[], c) => {
    if (hasImportDefaultSpecifierStyle(c)) {
      const specifiers = (c as ImportDeclaration).specifiers.filter(s => {
        return swc.isImportDefaultSpecifier(s)
      }) as ImportDefaultSpecifier[]
      p = [ ...p, ...specifiers ]
    }
    return p
  }, [] as ImportDefaultSpecifier[])
}