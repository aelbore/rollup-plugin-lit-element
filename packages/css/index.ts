import type { Plugin } from 'rollup'
import type { Enforce, Options, ViteInject } from 'shared'

import { InlineTransformer, Transformer } from './transformer'
import { swcTransformer, cssfilter, filter  } from 'shared'

export * from './transformer'
 
export function InlineCss(options?: Options) {
  const { vite, enforce = 'post' } = options || {}
  const isViteInject = (vite as ViteInject)?.inject
  return {
    name: 'css',
    enforce,
    transform(code: string, id: string) {
      const tsJsFilter = filter('js', options?.exclude || [])
      if (!tsJsFilter(id)  || (typeof vite == 'boolean' && vite) || isViteInject) return
      return swcTransformer(code, id, {
        paths: options?.paths,
        plugins: [ (p) => new InlineTransformer().visitProgram(p) ]
      })
    }
  } as Plugin & { enforce?: Enforce }
}

export function Css(options?: Options) {
  const { vite, enforce = 'post' } = options || {}
  return {
    name: 'css',
    enforce,
    transform(code: string, id: string) {
      if (!cssfilter(id.split('?')[0]) || !vite) return
      return swcTransformer(code, id, {
        paths: options?.paths,
        plugins: [ (p) => new Transformer(options).visitProgram(p) ]
      })
    }
  } as Plugin & { enforce?: Enforce }
} 