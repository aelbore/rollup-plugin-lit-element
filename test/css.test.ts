
import { Transformer } from 'css'

import { swcTransformer } from 'shared'

describe('css', () => {

  it('should update export to css template literals', () => {
    const content = `
      import { createHotContext as __vite__createHotContext } from "/@vite/client";
      import.meta.hot = __vite__createHotContext("/src/hello-world.scss");
      import { updateStyle as __vite__updateStyle, removeStyle as __vite__removeStyle } from "/@vite/client";
      const __vite__id = "/Users/override/Desktop/workspace/github/vite-lit-stater/src/hello-world.scss";
      const __vite__css = ":host h1 { color: blue; border: 1px solid green; padding: 10px; }";
      __vite__updateStyle(__vite__id, __vite__css);
      import.meta.hot.accept();
      export default __vite__css;
      import.meta.hot.prune(()=>__vite__removeStyle(__vite__id));
    `

    const { code } = swcTransformer(content, './file.ts', {
      plugins: [ (p) => new Transformer({ importPackage: 'lit' }).visitProgram(p) ]
    })

    expect(code).toMatchSnapshot()
  })

  it('should update the css template literals when it is default arrow', () => {
    const content = `
      export default (() => ":host h1{color:#00f;border:1px solid green;padding:10px}")()
    `

    const { code } = swcTransformer(content, './file.ts', {
      plugins: [ (p) => new Transformer({ importPackage: 'lit' }).visitProgram(p) ]
    })

    expect(code).toMatchSnapshot()
  })

})