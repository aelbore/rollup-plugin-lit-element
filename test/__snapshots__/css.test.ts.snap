// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`css > should update export to css template literals 1`] = `
"import { css as cssTag } from 'lit';
import { createHotContext as __vite__createHotContext } from \\"/@vite/client\\";
import.meta.hot = __vite__createHotContext(\\"/src/hello-world.scss\\");
import { updateStyle as __vite__updateStyle, removeStyle as __vite__removeStyle } from \\"/@vite/client\\";
const __vite__id = \\"/Users/override/Desktop/workspace/github/vite-lit-stater/src/hello-world.scss\\";
const __vite__css = cssTag\`:host h1 { color: blue; border: 1px solid green; padding: 10px; }\`;
__vite__updateStyle(__vite__id, __vite__css);
import.meta.hot.accept();
export default __vite__css;
import.meta.hot.prune(()=>__vite__removeStyle(__vite__id));
"
`;

exports[`css > should update the css template literals when it is default arrow 1`] = `
"import { css as cssTag } from 'lit';
export default cssTag\`:host h1{color:#00f;border:1px solid green;padding:10px}\`;
"
`;
